File: kangaroo_business_app_sdk_android/android/src/main/kotlin/com/kangaroo/flutterplugin/android/features/strings/StringsHandler.kt
@file:Suppress("INLINE_FROM_HIGHER_PLATFORM")
@file:OptIn(ExperimentalJsExport::class)
package com.kangaroo.flutterplugin.android.features.strings

import com.kangaroo.flutterplugin.android.base.PluginChannelHandler
import com.kangaroo.flutterplugin.android.base.pushSerializedResultToEventSink
import features.strings.StringsApi
import features.strings.models.ApplicationStringsModel
import features.strings.serializeStringsState
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import kangaroorewards.appsdk.core.domain.SerializedResult
import kangaroorewards.appsdk.core.domain.toJsonResult
import kotlin.js.ExperimentalJsExport


class StringsHandler : EventChannel.StreamHandler, PluginChannelHandler {
    var sink: EventChannel.EventSink? = null

    override val methodChannel: String
        get() = "customer_sdk/methods/get_strings"

    override val eventChannel: String
        get() = "customer_sdk/events/get_strings"

    override suspend fun onMethodCall(call: MethodCall): String? {
        return getStrings(call)
    }

    override fun getStreamHandler(): EventChannel.StreamHandler {
        return this
    }

    companion object {
        suspend fun getStrings(call: MethodCall): String? {
            val result = StringsApi().getStrings().toJsonResult<ApplicationStringsModel>()

            return when (result) {
                is SerializedResult.Success -> result.data
                is SerializedResult.UnauthorizedError -> result.error
                is SerializedResult.UnknownError -> result.error
                else -> null
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
        StringsApi().stringsState.serializeStringsState().observe { result ->
            pushSerializedResultToEventSink<StringsApi>(sink, result)
        }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}

File: kangaroo_business_app_sdk_ios/ios/Classes/features/strings/StringsHandler.swift
import Foundation
import Flutter
import KangarooAppSdkBusiness

class StringsHandler: NSObject, FlutterStreamHandler, PluginChannelHandler {
    var sink: FlutterEventSink?

    var methodChannel: String = "customer_sdk/methods/get_strings"

    var eventChannel: String = "customer_sdk/events/get_strings"

    func onMethodCall(call: FlutterMethodCall) async -> Any? {
        return await StringsHandler.getStrings(call: call)
    }

    func getStreamHandler() -> (FlutterStreamHandler & NSObjectProtocol)? {
        return self
    }


    static func getStrings(call: FlutterMethodCall) async -> String? {
        do {
        let result = try await StringsApi().getStrings().serializeStringsApiResult()

        switch result {
            case let result as SerializedResultSuccess:
                return result.data
            case let result as SerializedResultUnauthorizedError:
                return result.error
            case let result as SerializedResultUnknownError:
                return result.error
            default:
                return nil
            }
        } catch {
            return nil
        }

        


        

        
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping
        FlutterEventSink) -> FlutterError? {
        sink = events
        StringsApi().stringsState
            .serializeStringsState().observe { result in
            switch result {
            case let result as SerializedResultIdle:
                self.sink?(result.state)
            case let result as SerializedResultLoading:
                self.sink?(result.state)
            case let result as SerializedResultSuccess:
                self.sink?(result.data)
            case let result as SerializedResultEmptyResponse:
                self.sink?(result.body)
            case let result as SerializedResultUnauthorizedError:
                self.sink?(result.error)
            case let result as SerializedResultUnknownError:
                self.sink?(result.error)
            default:
                break
            }
        }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
